### ISSUE TITLE
Settings should reflect the configuration of the test suite.

### ISSUE THREAD

**Issue by richtja** [avocado-framework/avocado#5257]
Right now in the avocado, there is singleton `settings` which should holds configuration values of current avocado run. This is working fine until the tests are spited into different test suites with different config, or we are running avocado through Job API. In such cases `settings` contains only default configuration, which is not related to the test suite. 

Because of this problem, we can't use `settings` under suite context, and we have to pass config from suite into every method which is using configuration (this is our current solution) even the configuration is static for the suite context. Also, using `settings.as_dict()` might be problematic, because the result might be different from the current configuration. So we have to be sure that `settings.as_dict()` is used outside of the suite context.

In many places, we are using `settings.as_dict()` as substitution when the configuration wasn't passed, but with such solution we are hiding the problem, because such solution will be working, but the results will be wrong when the configurations differ. For example:

https://github.com/avocado-framework/avocado/blob/master/avocado/core/spawners/common.py#L33
https://github.com/avocado-framework/avocado/blob/master/avocado/core/test.py#L272
https://github.com/avocado-framework/avocado/blob/master/avocado/core/plugin_interfaces.py#L224

I would like to propose to change settings base on the given suite. Then there will be no need to pass config variable everywhere, but we could use `settings.as_dict()` instead and the `settings` will be responsible for providing the correct configuration. With such solution, we will unite approach to configuration for the whole avocado.

**Comment by beraldoleal** [comment]
> Right now in the avocado, there is singleton `settings` which should holds configuration values of current avocado run. This is working fine until the tests are spited into different test suites with different config, or we are running avocado through Job API. In such cases `settings` contains only default configuration, which is not related to the test suite.

Can you give us any example here? I tried to run: `python3 examples/jobs/multiple_suites_from_config.py` (which is using Job API and multiple suites) but I'm not sure if I'm following the what is not "working fine".

> Because of this problem, we can't use `settings` under suite context, and we have to pass config from suite into every method which is using configuration (this is our current solution) even the configuration is static for the suite context. Also, using `settings.as_dict()` might be problematic, because the result might be different from the current configuration. So we have to be sure that `settings.as_dict()` is used outside of the suite context.
> 
> In many places, we are using `settings.as_dict()` as substitution when the configuration wasn't passed, but with such solution we are hiding the problem, because such solution will be working, but the results will be wrong when the configurations differ. For example:
> 
> https://github.com/avocado-framework/avocado/blob/master/avocado/core/spawners/common.py#L33 https://github.com/avocado-framework/avocado/blob/master/avocado/core/test.py#L272 https://github.com/avocado-framework/avocado/blob/master/avocado/core/plugin_interfaces.py#L224

Those are examples that are using the as_dict() only when the configuration is empty. Probably I'm still missing something, could elaborate here? What do you mean with "configurations differ" ? as_dict() is returning the parsed settings respecting the order (default -> config file -> command-line options).

> I would like to propose to change settings base on the given suite. Then there will be no need to pass config variable everywhere, but we could use `settings.as_dict()` instead and the `settings` will be responsible for providing the correct configuration. With such solution, we will unite approach to configuration for the whole avocado.

I'm assuming you mean "based".

I'm ok with merging the methods to access the settings, I just would like to understand better the current problem, probably I'm missing something. So if you could give me some example codes I would appreciate.

**Comment by richtja** [comment]
> > Right now in the avocado, there is singleton `settings` which should holds configuration values of current avocado run. This is working fine until the tests are spited into different test suites with different config, or we are running avocado through Job API. In such cases `settings` contains only default configuration, which is not related to the test suite.
> 
> Can you give us any example here? I tried to run: `python3 examples/jobs/multiple_suites_from_config.py` (which is using Job API and multiple suites) but I'm not sure if I'm following the what is not "working fine".

Yes, the avocado doesn't have any issues with this example, because, we avoid calling the `settings` in suite context, instead we are passing the suite.config dict to every method where the configuration is needed. I am not talking about this as about bug. But I think, this approach needs refactor, so instead of passing the config everywhere we can use `settings` to provide the configuration. 

> 
> > Because of this problem, we can't use `settings` under suite context, and we have to pass config from suite into every method which is using configuration (this is our current solution) even the configuration is static for the suite context. Also, using `settings.as_dict()` might be problematic, because the result might be different from the current configuration. So we have to be sure that `settings.as_dict()` is used outside of the suite context.
> > In many places, we are using `settings.as_dict()` as substitution when the configuration wasn't passed, but with such solution we are hiding the problem, because such solution will be working, but the results will be wrong when the configurations differ. For example:
> > https://github.com/avocado-framework/avocado/blob/master/avocado/core/spawners/common.py#L33 https://github.com/avocado-framework/avocado/blob/master/avocado/core/test.py#L272 https://github.com/avocado-framework/avocado/blob/master/avocado/core/plugin_interfaces.py#L224
> 
> Those are examples that are using the as_dict() only when the configuration is empty. Probably I'm still missing something, could elaborate here? What do you mean with "configurations differ" ? as_dict() is returning the parsed settings respecting the order (default -> config file -> command-line options).

Yes, the as_dict() is only used when the configuration is empty. But any of those methods can't be used without test suite, so what is the use-case to have configuration empty or not related to the suite. IMO such a use-case doesn't exist, which means these examples are wrong.

I would like to have settings respecting such order: default -> config file -> command-line options or Job API configuration

> 
> > I would like to propose to change settings base on the given suite. Then there will be no need to pass config variable everywhere, but we could use `settings.as_dict()` instead and the `settings` will be responsible for providing the correct configuration. With such solution, we will unite approach to configuration for the whole avocado.
> 
> I'm assuming you mean "based".

Yes

> 
> I'm ok with merging the methods to access the settings, I just would like to understand better the current problem, probably I'm missing something. So if you could give me some example codes I would appreciate.

For example, this is the creation of status server inside nrunner:

```
    def _determine_status_server_uri(self, test_suite):
        # pylint: disable=W0201
        self.status_server_dir = None
        if test_suite.config.get('nrunner.status_server_auto'):
            # no UNIX domain sockets on Windows
            if platform.system() != 'Windows':
                self.status_server_dir = tempfile.TemporaryDirectory(
                    prefix='avocado_')
                return os.path.join(self.status_server_dir.name,
                                    '.status_server.sock')
        return test_suite.config.get('nrunner.status_server_listen')

    def _create_status_server(self, test_suite, job):
        listen = self._determine_status_server_uri(test_suite)
        # pylint: disable=W0201
        self.status_repo = StatusRepo(job.unique_id)
        # pylint: disable=W0201
        self.status_server = StatusServer(listen,
                                          self.status_repo)

```
There is no need to pass `test_suite` object, because the methods is working only with config. Also, the config is static in the context of suite, so IMO we don't need to pass the config either. I would like to be able to do this:
```diff
-    def _determine_status_server_uri(self, test_suite):
+    def _determine_status_server_uri(self):
         # pylint: disable=W0201
         self.status_server_dir = None
-        if test_suite.config.get('nrunner.status_server_auto'):
+        config = settings.as_dict()
+        if config.get('nrunner.status_server_auto'):
             # no UNIX domain sockets on Windows
             if platform.system() != 'Windows':
                 self.status_server_dir = tempfile.TemporaryDirectory(
                     prefix='avocado_')
                 return os.path.join(self.status_server_dir.name,
                                     '.status_server.sock')
-        return test_suite.config.get('nrunner.status_server_listen')
+        return config.get('nrunner.status_server_listen')
 
-    def _create_status_server(self, test_suite, job):
-        listen = self._determine_status_server_uri(test_suite)
+    def _create_status_server(self, job):
+        listen = self._determine_status_server_uri()
         # pylint: disable=W0201
         self.status_repo = StatusRepo(job.unique_id)
         # pylint: disable=W0201
```

**Comment by beraldoleal** [comment]
So, I talked with @richtja today about this thread. And we are on the same page. I will try my best to summarize what we discussed:

Currently, we don't have any significant issues or bugs here. This is more like a design choice, but we still have some things that could be improved to help the understanding. Because of the way this settings mechanism was designed, we are trying our best to split the "domains" by using the dictionary keys namespaces (example: job.* vs plugins.*), which would avoid having multiple configs. (Yes, we still have a lot of room for improvements here, especially on the namespaces).

Avocado has a distributed architecture, and sometimes when in an isolated environment different from the tier 0 bootstrapped environment, `settings.as_dict()` it will not return the proper configuration, so it cannot be used in all places. We *need* a communication channel to pass the right dictionary. And right now, we are using the method arguments as "the channel",  but if we need to remove those arguments, we need to implement a different communication channel, as the PUB/SUB mechanism, [already discussed here](https://github.com/avocado-framework/avocado/discussions/4316).

Here is what we could improve:

1. settings.as_dict() singleton:  give developers impression that could be called from everywhere. We should call it only as last resource, if the code you are working on it is not receiving the config.
   * AI: Better documentation of this;
   * AI: Try to remove current calls from settings.as_dict() (singleton), but maybe some of those are not possible (to investigate). If all are removed then we could deprecate the singleton (but keep as_dict() for convenience).
2. TestSuite.from_config(job_config): When using the Job API, currently we are *assuming* this is a config dictionary with only "job.*" namespaces. So if someone adds a different namespace here it will override default settings, creating inconsistency. We need to better document this and maybe validate the namespaces here.

@richtja, thanks again for the patience for both, waiting a reply and explaining in details your concerns. Let me know if I got something wrong, or something is still concerning you. And I guess, if you agree, we can close this issue.
